include/deal.II/dofs/dof_accessor.templates.h

```diff
@@ -197,6 +197,49 @@ namespace internal
 {
   namespace DoFAccessorImplementation
   {
+    /**
+     * Convert an FE index that might contain the right value but also
+     * invalid_fe_index to a right value if needed/possible.
+     */
+    template <int structdim, int dim, int spacedim, bool level_dof_access>
+    unsigned int
+    get_fe_index(
+      const DoFAccessor<structdim, dim, spacedim, level_dof_access> &cell,
+      const unsigned int                                             fe_index)
+    {
+      if (cell.get_dof_handler().has_hp_capabilities() == false)
+        {
+          // No hp enabled, and the argument is at its default value -> we
+          // can translate to the default active fe index
+          Assert(
+            (fe_index == DoFHandler<dim, spacedim>::invalid_fe_index) ||
+              (fe_index == DoFHandler<dim, spacedim>::default_fe_index),
+            ExcMessage(
+              "It is not possible to specify a FE index if no hp support is used!"));
+
+          return DoFHandler<dim, spacedim>::default_fe_index;
+        }
+      else
+        {
+          // Otherwise: If anything other than the default is provided by
+          // the caller, then we should take just that. As an exception, if
+          // we are on a cell (rather than a face/edge/vertex), then we know
+          // that there is only one active fe index on this cell and we can
+          // use that:
+          if ((dim == structdim) &&
+              (fe_index == DoFHandler<dim, spacedim>::invalid_fe_index))
+            {
+              return cell.nth_active_fe_index(0);
+            }
```

Comment:
Maybe we can be extra cautious here.
```suggestion
          if ((dim == structdim) &&
              (fe_index == DoFHandler<dim, spacedim>::invalid_fe_index))
            {
              Assert(cell.n_active_fe_indices() == 1, ExcInternalError());
              return cell.nth_active_fe_index(0);
            }
```
or:
```suggestion
          if ((cell.n_active_fe_indices() == 1) &&
              (fe_index == DoFHandler<dim, spacedim>::invalid_fe_index))
            {
              return cell.nth_active_fe_index(0);
            }
```
or my favorite:
```suggestion
          if (cell.n_active_fe_indices() == 1)
            {
              return cell.nth_active_fe_index(0);
            }
```

What do you think?

Reply:
I like option 2. Option three would need an assertion that `fe_index` either equals `invalid_fe_index` or the correct index.

Reply:
I went with option 1. This is an optimization for cells so let's keep it this way.